linker fails to link debug build due to missing symbols (including JS_SetGCZeal)

After configuring with `--debug` and `--with-external-spidermonkey-debug`:

```bash
> ./configure --engine=spidermonkey --debug --with-external-spidermonkey-release=/Users/myk/Projects/gecko-dev/obj-spidermonkey-for-spidershim --with-external-spidermonkey-debug=/Users/myk/Projects/gecko-dev/obj-spidermonkey-for-spidershim-debug --debug
```

The linker fails on errors when linking the debug build, including (but not limited to) _JS_SetGCZeal_ being undefined:

> make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C out BUILDTYPE=Release V=1
make[1]: Nothing to be done for `all'.
ln -fs out/Release/node node
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C out BUILDTYPE=Debug V=1
  c++ -Wl,-no_pie -Wl,-search_paths_first /Users/myk/Projects/spidernode/out/Debug/spidermonkey/Debug/icudata.o -lnspr4 -mmacosx-version-min=10.9 -arch x86_64 -L/Users/myk/Projects/spidernode/out/Debug -L/Users/myk/Projects/spidernode/out/Debug -L/Users/myk/Projects/spidernode/out/Debug/spidermonkey/Debug  -o "/Users/myk/Projects/spidernode/out/Debug/arraybuffer" /Users/myk/Projects/spidernode/out/Debug/obj.target/arraybuffer/deps/spidershim/test/arraybuffer.o /Users/myk/Projects/spidernode/out/Debug/libspidershim.a /Users/myk/Projects/spidernode/out/Debug/libgtest.a /Users/myk/Projects/spidernode/out/Debug/libzlib.a -ljs_static -lz -lmozglue -lspidershim
ld: warning: could not create compact unwind for _ffi_call_unix64: does not use RBP or RSP based frame
Undefined symbols for architecture x86_64:
  "JS_SetGCZeal(JSContext*, unsigned char, unsigned int)", referenced from:
      v8::Context::CreateGlobal(JSContext*, v8::Isolate*, v8::Local<v8::ObjectTemplate>) in libspidershim.a(v8context.o)
  "JSAutoCompartment::JSAutoCompartment(JSContext*, JSObject*, mozilla::detail::GuardObjectNotifier&&)", referenced from:
      v8::Object::Set(v8::Local<v8::Context>, v8::Local<v8::Value>, v8::Local<v8::Value>, v8::PropertyAttribute, bool) in libspidershim.a(v8object.o)
      v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>) in libspidershim.a(v8object.o)
      v8::Context::Impl::RunMicrotasks() in libspidershim.a(v8context.o)
      v8::Isolate::GetHiddenGlobal() in libspidershim.a(v8isolate.o)
      v8::ObjectTemplate::NewInstance(v8::Local<v8::Object>, v8::ObjectTemplate::ObjectType) in libspidershim.a(v8objecttemplate.o)
      void mozilla::Maybe<JSAutoCompartment>::emplace<JSContext*&, JS::Rooted<JSObject*>&>(JSContext*&&&, JS::Rooted<JSObject*>&&&) in libspidershim.a(v8objecttemplate.o)
  "JSAutoCompartment::JSAutoCompartment(JSContext*, JSObject*, mozilla::detail::GuardObjectNotifier&&)", referenced from:
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8typedarray.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8arraybuffer.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8arraybufferview.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8object.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8script.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8string.o)
      AutoJSAPI::AutoJSAPI(JSContext*, JSObject*) in libspidershim.a(v8context.o)
      ...
  "JS::AutoAssertOnGC::AutoAssertOnGC()", referenced from:
      JS::AutoCheckCannotGC::AutoCheckCannotGC() in libspidershim.a(v8string.o)
  "JS::AutoAssertOnGC::~AutoAssertOnGC()", referenced from:
      JS::AutoCheckCannotGC::~AutoCheckCannotGC() in libspidershim.a(v8string.o)
  "JS::detail::AssertArgumentsAreSane(JSContext*, JS::Handle<JS::Value>)", referenced from:
      JS::ToNumber(JSContext*, JS::Handle<JS::Value>, double*) in libspidershim.a(v8value.o)
      JS::ToInt32(JSContext*, JS::Handle<JS::Value>, int*) in libspidershim.a(v8value.o)
      JS::ToUint32(JSContext*, JS::Handle<JS::Value>, unsigned int*) in libspidershim.a(v8value.o)
      JS::ToString(JSContext*, JS::Handle<JS::Value>) in libspidershim.a(v8value.o)
      JS::ToObject(JSContext*, JS::Handle<JS::Value>) in libspidershim.a(v8value.o)
      JS::ToString(JSContext*, JS::Handle<JS::Value>) in libspidershim.a(v8function.o)
      JS::ToObject(JSContext*, JS::Handle<JS::Value>) in libspidershim.a(v8function.o)
      ...
  "js::oom::failAlways", referenced from:
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8object.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8string.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8context.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8isolate.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8objecttemplate.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8function.o)
  "js::oom::targetThread", referenced from:
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8object.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8string.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8context.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8isolate.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8objecttemplate.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8function.o)
  "js::oom::GetThreadType()", referenced from:
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8object.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8string.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8context.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8isolate.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8objecttemplate.o)
      js::oom::IsThreadSimulatingOOM() in libspidershim.a(v8function.o)
  "js::oom::maxAllocations", referenced from:
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8object.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8string.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8context.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8isolate.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8objecttemplate.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8function.o)
  "js::oom::counter", referenced from:
      js::oom::ShouldFailWithOOM() in libspidershim.a(v8object.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8object.o)
      js::oom::ShouldFailWithOOM() in libspidershim.a(v8string.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8string.o)
      js::oom::ShouldFailWithOOM() in libspidershim.a(v8context.o)
      js::oom::IsSimulatedOOMAllocation() in libspidershim.a(v8context.o)
      js::oom::ShouldFailWithOOM() in libspidershim.a(v8isolate.o)
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [/Users/myk/Projects/spidernode/out/Debug/arraybuffer] Error 1
make: *** [node_g] Error 2
09-06 17:37 > 

Over in https://github.com/mozilla/spidernode/pull/193#issuecomment-245130580, @ehsan thought this was unrelated to the bug in spidermonkey-external.gyp that copies release binaries to the debug directory, and he asked me to file a separate issue, hence this issue. But in https://github.com/mozilla/spidernode/pull/193#issuecomment-245135188, I explained why this is indeed caused by that bug, which #193 fixes.

--- a/configure
+++ b/configure
@@ -967,6 +967,8 @@ def configure_spidermonkey(o):
       raise Exception('--with-external-spidermonkey-debug passed without specifying --with-external-spidermonkey-release')
   if sm_debug != "":
     o['variables']['external_spidermonkey_debug'] = options.external_spidermonkey_debug
+  else:
+    o['variables']['external_spidermonkey_debug'] = options.external_spidermonkey_release
   o['variables']['external_spidermonkey_release'] = options.external_spidermonkey_release
 
 

--- a/deps/spidershim/spidermonkey-external.gyp
+++ b/deps/spidershim/spidermonkey-external.gyp
@@ -112,6 +112,7 @@
               ['external_spidermonkey_release_has_nspr == 1', {
                 # Normally we'd use libraries here, but gyp doesn't allow us.
                 'ldflags': [ '-lnspr4' ],
+                'xcode_settings': {'OTHER_LDFLAGS': ['-lnspr4']},
               }],
             ],
           },
@@ -125,6 +126,7 @@
               ['external_spidermonkey_debug_has_nspr == 1', {
                 # Normally we'd use libraries here, but gyp doesn't allow us.
                 'ldflags': [ '-lnspr4' ],
+                'xcode_settings': {'OTHER_LDFLAGS': ['-lnspr4']},
               }],
             ],
           },

--- a/deps/spidershim/spidermonkey-external.gyp
+++ b/deps/spidershim/spidermonkey-external.gyp
@@ -1,9 +1,9 @@
 {
   'variables': {
-    'external_spidermonkey_debug%': '<(external_spidermonkey_release)',
+    'external_spidermonkey_debug': '<(external_spidermonkey_release)',
     'spidermonkey_binaries_debug': [
-      '<(external_spidermonkey_debug%)/js/src/<(STATIC_LIB_PREFIX)js_static<(STATIC_LIB_SUFFIX)',
-      '<(external_spidermonkey_debug%)/config/external/icu/data/icudata.o',
+      '<(external_spidermonkey_debug)/js/src/<(STATIC_LIB_PREFIX)js_static<(STATIC_LIB_SUFFIX)',
+      '<(external_spidermonkey_debug)/config/external/icu/data/icudata.o',
     ],
     'spidermonkey_binaries_release': [
       '<(external_spidermonkey_release)/js/src/<(STATIC_LIB_PREFIX)js_static<(STATIC_LIB_SUFFIX)',
@@ -12,7 +12,7 @@
     'conditions': [
       ['external_spidermonkey_debug_has_nss == 1', {
         'spidermonkey_binaries_debug': [
-          '<(external_spidermonkey_debug%)/dist/lib/<(SHARED_LIB_PREFIX)nss3<(SHARED_LIB_SUFFIX)',
+          '<(external_spidermonkey_debug)/dist/lib/<(SHARED_LIB_PREFIX)nss3<(SHARED_LIB_SUFFIX)',
         ],
       }],
       ['external_spidermonkey_release_has_nss == 1', {
@@ -22,7 +22,7 @@
       }],
       ['OS == "linux"', {
         'spidermonkey_binaries_debug+': [
-          '<(external_spidermonkey_debug%)/mozglue/build/<(STATIC_LIB_PREFIX)mozglue<(STATIC_LIB_SUFFIX)',
+          '<(external_spidermonkey_debug)/mozglue/build/<(STATIC_LIB_PREFIX)mozglue<(STATIC_LIB_SUFFIX)',
         ],
         'spidermonkey_binaries_release+': [
           '<(external_spidermonkey_release)/mozglue/build/<(STATIC_LIB_PREFIX)mozglue<(STATIC_LIB_SUFFIX)',
@@ -30,7 +30,7 @@
       }],
       ['OS == "linux" and external_spidermonkey_debug_has_nspr == 1', {
         'spidermonkey_binaries_debug+': [
-          '<(external_spidermonkey_debug%)/config/external/nspr/pr/<(SHARED_LIB_PREFIX)nspr4<(SHARED_LIB_SUFFIX)',
+          '<(external_spidermonkey_debug)/config/external/nspr/pr/<(SHARED_LIB_PREFIX)nspr4<(SHARED_LIB_SUFFIX)',
         ],
       }],
       ['OS == "linux" and external_spidermonkey_release_has_nspr == 1', {
@@ -40,7 +40,7 @@
       }],
       ['OS == "mac"', {
         'spidermonkey_binaries_debug': [
-          '<(external_spidermonkey_debug%)/dist/bin/<(SHARED_LIB_PREFIX)mozglue<(SHARED_LIB_SUFFIX)',
+          '<(external_spidermonkey_debug)/dist/bin/<(SHARED_LIB_PREFIX)mozglue<(SHARED_LIB_SUFFIX)',
         ],
         'spidermonkey_binaries_release': [
           '<(external_spidermonkey_release)/dist/bin/<(SHARED_LIB_PREFIX)mozglue<(SHARED_LIB_SUFFIX)',
@@ -48,7 +48,7 @@
       }],
       ['OS == "mac" and external_spidermonkey_debug_has_nspr == 1', {
         'spidermonkey_binaries_debug': [
-          '<(external_spidermonkey_debug%)/dist/lib/<(SHARED_LIB_PREFIX)nspr4<(SHARED_LIB_SUFFIX)',
+          '<(external_spidermonkey_debug)/dist/lib/<(SHARED_LIB_PREFIX)nspr4<(SHARED_LIB_SUFFIX)',
         ],
       }],
       ['OS == "mac" and external_spidermonkey_release_has_nspr == 1', {

--- a/deps/spidershim/spidermonkey-external.gyp
+++ b/deps/spidershim/spidermonkey-external.gyp
@@ -1,6 +1,5 @@
 {
   'variables': {
-    'external_spidermonkey_debug': '<(external_spidermonkey_release)',
     'spidermonkey_binaries_debug': [
       '<(external_spidermonkey_debug)/js/src/<(STATIC_LIB_PREFIX)js_static<(STATIC_LIB_SUFFIX)',
       '<(external_spidermonkey_debug)/config/external/icu/data/icudata.o',

